llvm/lib/Target/SPIRV/SPIRVEmitNonSemanticDI.cpp

```diff
@@ -0,0 +1,187 @@
+
+char SPIRVEmitNonSemanticDI::ID = 0;
+
+SPIRVEmitNonSemanticDI::SPIRVEmitNonSemanticDI(SPIRVTargetMachine *TM)
+    : MachineFunctionPass(ID), TM(TM) {
+  initializeSPIRVEmitNonSemanticDIPass(*PassRegistry::getPassRegistry());
+}
+
+SPIRVEmitNonSemanticDI::SPIRVEmitNonSemanticDI() : MachineFunctionPass(ID) {
+  initializeSPIRVEmitNonSemanticDIPass(*PassRegistry::getPassRegistry());
+}
+
+bool SPIRVEmitNonSemanticDI::emitGlobalDI(MachineFunction &MF) {
+  // If this MachineFunction doesn't have any BB repeat procedure
+  // for the next
+  if (MF.begin() == MF.end()) {
+    IsGlobalDIEmitted = false;
+    return false;
+  }
+
+  // Required variables to get from metadata search
+  LLVMContext *Context;
+  SmallString<128> FilePath;
+  unsigned SourceLanguage = 0;
+  int64_t DwarfVersion = 0;
+  int64_t DebugInfoVersion = 0;
+
+  // Searching through the Module metadata to find nescessary
+  // information like DwarfVersion or SourceLanguage
+  {
+    const MachineModuleInfo &MMI =
+        getAnalysis<MachineModuleInfoWrapperPass>().getMMI();
+    const Module *M = MMI.getModule();
+    Context = &M->getContext();
+    const NamedMDNode *DbgCu = M->getNamedMetadata("llvm.dbg.cu");
+    if (!DbgCu)
+      return false;
+    for (const auto *Op : DbgCu->operands()) {
+      if (const auto *CompileUnit = dyn_cast<DICompileUnit>(Op)) {
+        DIFile *File = CompileUnit->getFile();
+        sys::path::append(FilePath, File->getDirectory(), File->getFilename());
+        SourceLanguage = CompileUnit->getSourceLanguage();
+        break;
+      }
+    }
+    const NamedMDNode *ModuleFlags = M->getNamedMetadata("llvm.module.flags");
+    for (const auto *Op : ModuleFlags->operands()) {
+      const MDOperand &MaybeStrOp = Op->getOperand(1);
+      if (MaybeStrOp.equalsStr("Dwarf Version"))
+        DwarfVersion =
+            cast<ConstantInt>(
+                cast<ConstantAsMetadata>(Op->getOperand(2))->getValue())
+                ->getSExtValue();
+      else if (MaybeStrOp.equalsStr("Debug Info Version"))
+        DebugInfoVersion =
+            cast<ConstantInt>(
+                cast<ConstantAsMetadata>(Op->getOperand(2))->getValue())
+                ->getSExtValue();
+    }
+  }
+  // NonSemantic.Shader.DebugInfo.100 global DI instruction emitting
+  {
+    // Required LLVM variables for emitting logic
+    const SPIRVInstrInfo *TII = TM->getSubtargetImpl()->getInstrInfo();
+    const SPIRVRegisterInfo *TRI = TM->getSubtargetImpl()->getRegisterInfo();
+    const RegisterBankInfo *RBI = TM->getSubtargetImpl()->getRegBankInfo();
+    SPIRVGlobalRegistry *GR = TM->getSubtargetImpl()->getSPIRVGlobalRegistry();
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    MachineBasicBlock &MBB = *MF.begin();
+
+    // To correct placement of a OpLabel instruction during SPIRVAsmPrinter
+    // emission all new instructions needs to be placed after OpFunction
+    MachineIRBuilder MIRBuilder(MBB, MBB.end());
```

Comment:
@bwlodarcz I'm afraid this approach doesn't work. When you insert new instructions at the end of the machine basic block, after a function, they appear *after* a terminator instruction that is considered to be an error from the perspective of Machine verifier. Please run the test case from this PR with `-verify-machineinstrs` (`llc -verify-machineinstrs --spv-emit-nonsemantic-debug-info --spirv-ext=+SPV_KHR_non_semantic_info ...`) to see a list of errors. As an example, they will look like:
```
*** Bad machine code: Non-terminator instruction after the first terminator ***
- function:    foo
- basic block: %bb.1 entry (0xb1c9b8)
- instruction: %8:iid(s32) = OpConstantI %5:type(s64), 21
First terminator was:   OpReturn
```

This means that this PR at the moment introduces a test case that fails under LLVM_ENABLE_EXPENSIVE_CHECKS=ON.
cc: @michalpaszkowski

Reply:
Just as an illustration of the issue, the output code from the test case:
```
# After SPIRV NonSemantic.Shader.DebugInfo.100 emitter
# Machine code for function foo: IsSSA, TracksLiveness, Legalized, Selected

bb.1.entry:
  %1:type = OpTypeVoid
  %2:type = OpTypeFunction %1:type
  %0:iid = OpFunction %1:type, 0, %2:type
  OpName %0:iid, 7303014
  OpDecorate %0:iid, 41, 7303014, 0
  OpReturn
  %3:id(s32) = OpString 1094795567, 1094795585, 792805697, 1111638594, 1111638594, 1128481583, 1128481603, 1697596227, 1886216568, 1663985004, 0
  %4:id(s32) = OpExtInst %1:type, 3, 35, %3:id(s32)
  %5:type(s64) = OpTypeInt 32, 0
  %6:iid(s32) = OpConstantI %5:type(s64), 5
  %7:iid(s32) = OpConstantI %5:type(s64), 3
  %8:iid(s32) = OpConstantI %5:type(s64), 21
  %9:id(s32) = OpExtInst %1:type, 3, 1, %7:iid(s32), %6:iid(s32), %4:id(s32), %8:iid(s32)

# End machine code for function foo.
```